onnxruntime/core/providers/cpu/llm/attention.cc

```diff
@@ -0,0 +1,749 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/llm/attention.h"
+
+#include "core/common/common.h"
+#include "core/common/safeint.h"
+#include "core/mlas/inc/mlas.h"
+#include "core/platform/threadpool.h"
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+
+using onnxruntime::attention_helper::AttentionMaskType;
+using onnxruntime::attention_helper::AttentionParameters;
+using onnxruntime::attention_helper::AttentionType;
+using onnxruntime::attention_helper::QKMatMulOutputMode;
+using onnxruntime::concurrency::ThreadPool;
+
+namespace onnxruntime {
+
+#define REGISTER_ONNX_KERNEL_TYPED(T)                                 \
+  ONNX_CPU_OPERATOR_TYPED_KERNEL(                                     \
+      Attention,                                                      \
+      23,                                                             \
+      T,                                                              \
+      KernelDefBuilder()                                              \
+          .TypeConstraint("T1", DataTypeImpl::GetTensorType<T>())     \
+          .TypeConstraint("T2", DataTypeImpl::GetTensorType<T>())     \
+          .TypeConstraint("U", BuildKernelDefConstraints<bool, T>()), \
+      Attention<T>);
+
+REGISTER_ONNX_KERNEL_TYPED(float)
+
+template <typename T>
+inline void ComputeAttentionSoftmaxInplace(T* score, int N, int D, ThreadPool* tp) {
+  MlasComputeSoftmax(score, score, N, D, false, false, tp);
+}
+
+template <typename T>
+inline void ComputeAttentionSoftcapInplace(T* scores, int sequence_length, T softcap) {
+  MlasComputeSoftcap(scores, scores, sequence_length, softcap);
+}
+
+template <typename T>
+Attention<T>::Attention(const OpKernelInfo& info) : AttentionBase<T>(info) {
+  is_causal_ = static_cast<int>(info.GetAttrOrDefault<int64_t>("is_causal", 0)) == 1;
+  kv_num_heads_ = static_cast<int>(info.GetAttrOrDefault<int64_t>("kv_num_heads", 0));
+  q_num_heads_ = static_cast<int>(info.GetAttrOrDefault<int64_t>("q_num_heads", 0));
+  int mode = static_cast<int>(info.GetAttrOrDefault<int64_t>("qk_matmul_output_mode", 0));
+  qk_matmul_output_mode_ = static_cast<QKMatMulOutputMode>(mode);
+  ORT_ENFORCE(qk_matmul_output_mode_ == QKMatMulOutputMode::kQK ||
+                  qk_matmul_output_mode_ == QKMatMulOutputMode::kQKMask ||
+                  qk_matmul_output_mode_ == QKMatMulOutputMode::kQKSoftCap ||
+                  qk_matmul_output_mode_ == QKMatMulOutputMode::kQKSoftMax,
+              "qk_matmul_output_mode must be 0, 1, 2, or 3.");
+  scale_ = info.GetAttrOrDefault<float>("scale", std::numeric_limits<T>::quiet_NaN());
```

Comment:
We can put a comment saying the default will be applied later in the implementation.

Reply:
done